### ISSUE TITLE
"Failed to zoom VST3" message and crash when opening with custom zoom value in Studio One Pro 5

### ISSUE THREAD

**Issue by anaveragebear** [surge-synthesizer/surge#2397]
**Describe the bug**
As stated in the title, after setting a default zoom value in Surge, my computer spits out the above message and my DAW crashes on attempted reopen.

**Please let us know your surge version**
1.7.0
VST3 // 64-bit

**To Reproduce**
1. Manually enter a default value in the Surge settings menu
2. Exit the program
3. Attempt to reopen the program
4. Bug

**Expected behavior**
An infinitely repeating text box containing the text in the title, GUI does not load, DAW freezes
![Screenshot (61)](https://user-images.githubusercontent.com/68977809/88869256-45abfc00-d1e0-11ea-8672-42d7d3e3333d.png)

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro v1909
 - Host: Studio One Pro 5.0.0.59360 64-bit

**Additional context**
I am a lowly music producer without a lick of coding knowledge but I love this synth and wanted to help improve it. Cheers

**Comment by baconpaul** [comment]
Thanks for the report!

@mkruselj do you have studio one? Can you see if you can reproduce this in the next few days?

@anaveragebear if you don't zoom does surge work? (I know that's not a solution but it is at least a workaround!). If not, you can always grab the previous release (1.6.6) from our website.

This error will appear once if your host doesn't respond with "True" from the vst3 "resize" message. I think it happens over and over because we've never actually seen that happen before so I bet we keep on trying. If we can reproduce it we can fix that loop case.

**Comment by baconpaul** [comment]
Oh and one more thing @anaveragebear - if you go to your %DOCUMENTS%\Surge directory there will be a file called "SurgeUserDefaults.xml". If you just delete that file, it will undo the default setting and you should be good to go at 100% again. (Which will be tiny but at least will work for now).

**Comment by anaveragebear** [comment]
Thanks @baconpaul I'll be sure to do that.

**Comment by mkruselj** [comment]
I do have S1 but v4 not v5...

**Comment by mkruselj** [comment]
In v4 I don't get the endless repeating dialog - but I do get the same message if you try to zoom outside of screen boundaries (when Surge warns you that won't work and it auto-reduces the zoom level to fit at largest possible size for the current resolution).

**Comment by mkruselj** [comment]
For reference - the same issue doesn't happen in Reaper - you get the dialog that zoom level is too large (confirming that the oversized zoom level actually **is** stored in plugin's chunk), but Reaper resizes it nicely and no further errors.

So I suppose the solution for this is after zoom level is readjusted if user overblown it, to write the adjusted value into the plugin chunk.

**Comment by baconpaul** [comment]
Yeah I think what is happening is if that zoom call fails we keep hammering away at the old level. And only S1 fails the call if you are too large. Very useful thanks.

I own pre-sonus gear so in theory I can get S1. I never have though. Lemme see if I can figure it out!

**Comment by mkruselj** [comment]
I very much think if the zoom level is overblown, it shouldn't be written in plugin's chunk. Then you get that zoom level warning dialog when you load your project, which is not nice.

**Comment by baconpaul** [comment]
Yeah so what would normally happen is, if you are overblown

1. you would get the error and 
2. surge would size down to a fit
3. that fit would then get written into the chunk

what seems to be happening in s1 is that surge is failing to size down so we get the loop with all the errors, because the VST3 resize system call fails (that's what pops that vst3 zoom failed error). 

I think the S1 window manager has its own "bigger than screen" check which fails me separately than my check. So I need to debug that code path.

**Comment by baconpaul** [comment]
Note to self from Evil and me talking on slack so I can record this here:

```
So what is happening is this.
12:14
S1 has a more aggressive ‘max size’ algo than surge
12:14
But S1 tells us that - unlike any other DAW - by having the VST3 resize API fail
12:15
That used to send us into a loop
12:15
Now it sends us back to 100%
12:15
But if you send back to 100% when you are trying to send back from 200 you end up one recursion depth further than I thought
12:15
So I think a fix which is not perfect but at least gets OP in the water is to up my recursion guard count by one, remove that error message, and then it will appear to work.
12:16
If you set 200, you will get surge telling you you are going to 190. Then S1 failing that and VST3 telling you it is going to 100.
12:16
See what I mean?
12:16
And only S1 fails that API (everyone else just resizes to larger than screen which is why I put a check in surge).
```

**Comment by baconpaul** [comment]
OK @anaveragebear Studio 1 is weird. We have a sort-of fix that I'll merge now and then you can grab our nightly build. It works with zoom particular zoom levels if you use the zoom menu but isn't perfect.

The remote debugging (where I debug while @mkruselj runs it) has gotten us pretty far but I really just need to get a copy of Studio One. I'll reach out to Presonus and see if they have a developer program or some such. Turns out if you buy a hardware interface from them they give you a copy of the DAW that doesn't load VSTs. But we can figure that out.

I'll update this issue again when the nightly is ready and perhaps you can give it a try?

**Comment by baconpaul** [comment]
@anaveragebear if you grab the nightly from https://surge-synthesier.github.io/nightly, the situation will be greatly improved. Not perfect but greatly improved. Some popups still but if you close and open should be OK.

**Comment by baconpaul** [comment]
@anaveragebear we released Surge 1.7.1 this weekend which contains the fix that @mkruselj and I came to. It isn't perfect but should get you back in the water in Studio One. If you get a chance to give it a whirl, please let us know.

**Comment by mkruselj** [comment]
OK so this should be fixed now for the current nightly (as of 8369a6f) and for 1.8 release! Closing issue.
